skcosmo/feature_selection/voronoi_fps.py

```diff
@@ -0,0 +1,136 @@
+import numpy as np
+
+from .simple_fps import SimpleFPS
+
+
+class VoronoiFPS(SimpleFPS):
+    """
+    Base Class defined for Voronoi FPS methods
+
+    :param initialize: predetermined index; if None provided, first index selected
+                 is 0
+    :type selected_: int, None
+    """
+
+    def __init__(self, voronoi_cutoff_fraction=1.0 / 6.0, **kwargs):
+        self.voronoi_cutoff_fraction = voronoi_cutoff_fraction
+        super().__init__(**kwargs)
+
+    def _init_greedy_search(self, X, y, n_to_select):
+
+        n_features = X.shape[1]
+
+        # self.voronoi_number should coincide with self.selected_idx_
+        # self.furthest_point = np.full(self.n_features_to_select, np.inf)
+
+        self.voronoi_location = np.full(n_features, 0)
+        self.idx_in_voronoi = [np.array([]) for n in range(self.n_features_to_select)]
+        self.idx_in_voronoi[0] = np.arange(n_features, dtype=int)
+        self.number_in_voronoi = np.full(self.n_features_to_select, 0)
+        self.number_in_voronoi[0] = n_features
+
+        # define the voronoi_r2 for the selected_ point
+        # this is the maximum distance from the center for each of the cells
+        # self.furthest_distance = np.zeros(self.n_features_to_select, float)
+
+        # quarter of the square distance between new selected point and previously
+        # selected points
+        self.sel_d2q_ = np.zeros(self.n_features_to_select, float)
+
+        super()._init_greedy_search(X, y, n_to_select)
+
+        # self.voronoi_r2[0] = self.haussdorf_[self.voronoi_i_far[0]]
+
+    def _get_active(self, X, last_selected):
+        # because we haven't called super() yet, self.n_selected_ has not been
+        # incremented yet
+
+        f_active = np.zeros(self.n_features_to_select, bool)
+
+        """must compute distance of the new point to all the previous FPS. Some
+            of these might have been computed already, but bookkeeping could be
+            worse that recomputing (TODO: verify!)
+            """
+        # calculation in a single block
+        self.sel_d2q_[: self.n_selected_] = (
+            self.norms_[self.selected_idx_[: self.n_selected_]]
+            + self.norms_[last_selected]
+            - 2 * (self.X_selected_[:, : self.n_selected_].T @ X[:, last_selected])
+        ) * 0.25
+
+        for ic in range(self.n_selected_):
+            # empty voronoi, no need to consider it
+            if self.number_in_voronoi[ic] > 1:
+                r2 = max(self.haussdorf_[self.idx_in_voronoi[ic]])
```

Comment:
We discussed with you that your main focus was to reduce the number of iterations of distance calculations. But still, iterating over components can seriously affect performance (in fact, cython was proposed for the sake of reducing the time to iterate). Here we are essentially checking all points in our space every iteration. Although this can be avoided by simply using the voronoi_r2 array. In the previous code its filling went along with moving points from one polyhedron to another, i.e. one more loop was avoided.  It seems to me that its use is still justified, or is it a mistake?

Reply:
```suggestion
                r2 = self.haussdorf_[self.furthest_point[ic]]
                
```